### PR TITLE
Command family cli subcommands

### PR DESCRIPTION
**Pull request by isabeltomb**
Adds subcommands to the transact cli to interact with the command family smart contract. This is behind the experimental feature "command".

### **Testing:**

1. Checkout the branch in this PR which updates the command family and adds a command family smart contract: https://github.com/hyperledger/transact/pull/126
2. Start two splinter nodes and create a circuit between them with scabbard services
3. Download the command family smart contract scar file found here: https://build.sawtooth.me/job/Transact-Hyperledger/job/transact/job/PR-127/1/artifact/build/scar/
4. Upload the command family smart contract to the circuit using the scabbard CLI
5. Create a namespace registry for the contract using the namespace "06abbc"
```
scabbard ns create 06abbc \
    --owner <public_key> \
    --key <path_to_private_key> \
    --url http://0.0.0.0:8080 \
    --service-id <circuit-id>::<scabbard-service-id>
```
6. Add read and write permissions for the contract
```
scabbard perm 06abbc command --read --write \
    --key <path_to_private_key> \
    --url http://0.0.0.0:8080 \
    --service-id <circuit-id>::<scabbard-service-id>
```
7. Start the transact CLI using the branch from step 1, if you are using a docker file for your splinter nodes you can add something similar to the following to your docker file: 
```
  transact-cli:
    image: transact-cli
    build:
      context: ../transact
      dockerfile: ./cli/Dockerfile
      args:
      - CARGO_ARGS=${CARGO_ARGS}
      - REPO_VERSION=0.4.1-dev
    container_name: transact-cli
    hostname: transact-cli
```
and then set the environment variable `CARGO_ARGS="-- --features=experimental"` and then run `docker-compose -f <docker-file-name>.yaml run transact-cli bash` in a new terminal window to start the transact CLI 
(alternatively the transact commands can be run from the transact/cli directory as `cargo run --features=experimental ...`)

8. Run a set-state command
```
transact command set-state \
  --key /alice.priv
  --target "http://0.0.0.0:8080/scabbard/<circuit-id>/<scabbard-service-id>"
  --state-entry 06abbcb16ed7d24b3ecbd4164dcdad374e08c0ab7518aa07f9d3683f34c2b3c67a1583:state-value
```
9. Check the splinterd logs to see if the transaction was successfully submitted
10. Run a show-state command
```
transact command show-state \
  --key /alice.priv
  --target "http://0.0.0.0:8080/scabbard/<circuit-id>/<scabbard-service-id>"
  --address 06abbcb16ed7d24b3ecbd4164dcdad374e08c0ab7518aa07f9d3683f34c2b3c67a1583
  --text
```
11. Check that the "state-value" is returned
12. Run a get-state command 
```
transact command get-state \
  --key /alice.priv
  --target "http://0.0.0.0:8080/scabbard/<circuit-id>/<scabbard-service-id>"
  --address 06abbcb16ed7d24b3ecbd4164dcdad374e08c0ab7518aa07f9d3683f34c2b3c67a1583
```
13. Check the splinterd logs to see if the get state transaction was successfully submitted, the transact cli should not display anything after running this command.

### CODE REVIEW

File: cli/man/transact-command-get-state.1.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+`-k, --key PRIVATE-KEY-FILE`
+: Specifies the full path to the private key file. The key will be used to
+  sign the batches as well as generate a JWT for authentication.
+
+`--target TARGET`
+: Node URL to submit batches to. The URL should include all of the information
+  required to append `/batches` to the end.
+
+`--address ADDRESS`
+: State address of the state to be read.
```
Comment (shannynalayna) [inline cli/man/transact-command-get-state.1.md:53]: Can multiple addresses be supplied here? If multiple may be supplied, would make sense to reference that fact here.  (Same goes for the show command if multiple addresses can be supplied)

File: cli/man/transact-command-get-state.1.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+
+EXAMPLES
+========
+The following shows submitting a get state transaction to a Splinter circuit
+`vpENT-eSfFZ` with scabbard services. A Scabbard service runs a Sabre
```
Comment (shannynalayna) [inline cli/man/transact-command-get-state.1.md:59]: Scabbard doesn't need to be capitalized, I believe. Should keep the capitalization consistent (this applies to the other man pages that mention scabbard)

**Review by shannynalayna — COMMENTED** [review]
I would also suggest splitting the last two commits based on the command being implemented

File: cli/src/action/command.rs [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+        let signature = signer
+            .sign(&header_bytes)
+            .map_err(|err| CliError::ActionError(format!("Unable to sign batch: {}", err)))?
+            .as_hex();
+        batch.set_header_signature(signature);
+        batch.set_header(header_bytes.to_vec());
+        batch.set_transactions(RepeatedField::from_vec(vec![txn]));
+
+        let batch_pair = BatchPair::from_proto(batch).map_err(|err| {
+            CliError::ActionError(format!("Unable to convert from proto: {}", err))
```
[the remark this answers is not in the document]
Comment (isabeltomb) [reply cli/src/action/command.rs]: Updated to include more details

File: cli/src/action/command.rs [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+                            CliError::ActionError(format!(
+                                "Set state request failed with status code '{}', but \
+                                    error response was not valid",
+                                status
+                            ))
+                        })?
+                        .message;
+
+                    Err(CliError::ActionError(format!(
+                        "Failed to submit: {}",
```
[the remark this answers is not in the document]
Comment (isabeltomb) [reply cli/src/action/command.rs]: Updated these error messages

File: cli/man/transact-command-get-state.1.md (hunk 1 above)
Comment (isabeltomb) [reply cli/man/transact-command-get-state.1.md]: Updated this to include info about multiple addresses

File: cli/man/transact-command-get-state.1.md (hunk 2 above)
Comment (isabeltomb) [reply cli/man/transact-command-get-state.1.md]: Updated all the man pages in this PR

**Review by shannynalayna — DISMISSED** [review]

**Review by agunde406 — APPROVED** [review]

**Review by shannynalayna — DISMISSED** [review]

File: cli/src/action/command.rs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+            state_writes.push(entry);
+
+            addresses.push(address);
+        }
+
+        let mut set_state = SetState::new();
+        set_state.set_state_writes(RepeatedField::from_vec(state_writes));
+
+        let mut command = Command::new();
```
Comment (peterschwarz) [inline cli/src/action/command.rs:69]: Non-blocking comment: This probably should have a builder/into_proto combination.

File: cli/man/transact-command-set-state.1.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+: Specifies the full path to the private key file. The key will be used to
+  sign the batches as well as generate a JWT for authentication.
+
+`--target TARGET`
+: Node URL to submit batches to. The URL should include all of the information
+  required to append `/batches` to the end.
+
+`--state-entry STATE-ENTRY`
+: Key-value pair where the key is a state address and the value is the value to
+  be set for that address. (format: key:value)
```
Comment (peterschwarz) [inline cli/man/transact-command-set-state.1.md:55]: s/key/address

Comment (isabeltomb) [reply cli/man/transact-command-set-state.1.md]: Fixed

**Review by peterschwarz — APPROVED** [review]

**Review by agunde406 — APPROVED** [review]

**Review by shannynalayna — APPROVED** [review]

**Pull request merged**